### PR TITLE
Made restarting the Engine remember the last entrypoint that was used.

### PR DESCRIPTION
**Pull request by gaaclarke**
Related issue: https://github.com/flutter/flutter/issues/33099

### CODE REVIEW

**Comment by gaaclarke** [comment]
> Is the use case that an isolate restart needs to take into the account the last entrypoint and library? If so, why shouldn't the asset manager be used as well? I would just make the engine cache the last run configuration. Then, callers can access it (via copy), modify it, and then use it to restart the engine.

I have concerns about keeping the Asset Manager alive longer than it may previously have been alive.  I did the minimal change the solves the problem.  Do you think it is safe to start keeping that object alive for as long as the Engine is alive?

**Comment by chinmaygarde** [comment]
> I have concerns about keeping the Asset Manager alive longer...

Can you elaborate on these concerns? If the isolate restart previously did not remember the custom entrypoint or library, isn't it a bug still that it forgets where its assets are after a restart?

**Comment by gaaclarke** [comment]
> Can you elaborate on these concerns? If the isolate restart previously did not remember the custom entrypoint or library, isn't it a bug still that it forgets where its assets are after a restart?

Sounds plausible, I've never seen that bug reported though.  That sounds like it would have been something someone noticed at some point.

**Comment by chinmaygarde** [comment]
That an isolate restart with custom entrypoints not remembering that custom entrypoint went unreported suggests this isn't a very thoroughly exercised workflow.

**Comment by gaaclarke** [comment]
Talked offline with @chinmaygarde.  It is still an open question if we want to save the AssetManager, because an AssetResolver is getting rebuilt when we reload:
https://github.com/gaaclarke/engine/blob/7caa83ef28f20d9f6b0b273202d4b04dda09cbfa/shell/common/shell.cc#L1237

I recommend we continue down this path of the simplest thing and then reevaluate if we want more.

File: shell/common/engine.h [hunk 1]
```diff
@@ -716,6 +716,20 @@ class Engine final : public RuntimeDelegate, PointerDataDispatcher::Delegate {
   // |PointerDataDispatcher::Delegate|
   void ScheduleSecondaryVsyncCallback(fml::closure callback) override;
 
+  //----------------------------------------------------------------------------
+  /// @brief      Get the last Entrypoint that was used in the RunConfiguration
+  ///             when |Engine::Run| was called.
+  ///
+  const std::string& GetEntrypoint() const { return entry_point_; }
```
Comment (chinmaygarde) [inline shell/common/engine.h]: `GetLastEntryPoint`. Also, please OOL this accessor.

File: shell/common/engine.h [hunk 2]
```diff
@@ -716,6 +716,20 @@ class Engine final : public RuntimeDelegate, PointerDataDispatcher::Delegate {
...
+
+  //----------------------------------------------------------------------------
+  /// @brief      Get the last Entrypoint Library that was used in the
+  ///             RunConfiguration when |Engine::Run| was called.
+  ///
+  const std::string& GetEntrypointLibrary() const {
```
Comment (chinmaygarde) [inline shell/common/engine.h]: `GetLastEntryPointLibrary`. Also, please OOL this accessor.

File: shell/common/engine.h [hunk 3]
```diff
@@ -727,6 +741,8 @@ class Engine final : public RuntimeDelegate, PointerDataDispatcher::Delegate {
   // is destructed first.
   std::unique_ptr<PointerDataDispatcher> pointer_data_dispatcher_;
 
+  std::string entry_point_;
```
Comment (chinmaygarde) [inline shell/common/engine.h]: `last_...`

File: shell/common/engine.h (hunk 1 above)
Comment (gaaclarke) [reply shell/common/engine.h]: done, we call them `Entrypoint`s everywhere not `EntryPoint`s

File: shell/common/engine.h (hunk 2 above)
Comment (gaaclarke) [reply shell/common/engine.h]: done

File: shell/common/engine.h (hunk 3 above)
Comment (gaaclarke) [reply shell/common/engine.h]: done

**Comment by gaaclarke** [comment]
@chinmaygarde Done as requested, added a test and made `Shell::GetEngine()` available to everyone while also filing a bug to create an alternative (https://github.com/flutter/flutter/issues/43224).

File: shell/common/engine.h [hunk 4]
```diff
@@ -716,6 +716,18 @@ class Engine final : public RuntimeDelegate, PointerDataDispatcher::Delegate {
...
+  /// @brief      Get the last Entrypoint that was used in the RunConfiguration
+  ///             when |Engine::Run| was called.
+  ///
+  const std::string& GetLastEntrypoint() const;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Get the last Entrypoint Library that was used in the
+  ///             RunConfiguration when |Engine::Run| was called.
+  ///
+  const std::string& GetLastEntrypointLibrary() const;
```
Comment (chinmaygarde) [inline shell/common/engine.h]: Very minor nit: Here and earlier. Missing the @return doxygen directive.

**Pull request merged**